webrender/src/device.rs

```diff
@@ -592,32 +594,50 @@ impl ProgramSources {
     }
 }
 
+#[cfg_attr(feature = "serialize_program", derive(Deserialize, Serialize))]
 pub struct ProgramBinary {
+    sources: ProgramSources,
     binary: Vec<u8>,
     format: gl::GLenum,
 }
 
 impl ProgramBinary {
-    fn new(binary: Vec<u8>, format: gl::GLenum) -> Self {
+    fn new(sources: &ProgramSources, binary: Vec<u8>, format: gl::GLenum) -> Self {
         ProgramBinary {
+            sources: sources.clone(),
             binary,
             format
         }
     }
 }
 
+/// The interfaces that an application can implement to handle ProgramCache update
+pub trait ProgramCacheObserver {
+    fn notify_binary_added(&self, program_binary: &Arc<ProgramBinary>);
```

Comment:
does the application need to have `Arc<>` here? would it make sense to provide `&ProgramBinary` instead?

Reply:
Arc<> is necessary here, since ProgramBinary is serialized and is stored to disk on worker thread.